flare/gp_from_aimd.py

```diff
@@ -252,18 +252,357 @@ def __init__(
+            self.gp.check_L_alpha()
+            logger.debug(f"Done check_L_alpha {time.time() - time0}")
+
+    def run_active_learning(
+        self,
+        frames: Union[List[Structure], Trajectory] = (),
+        rel_std_tolerance: float = 4,
+        abs_std_tolerance: float = 0,
+        abs_force_tolerance: float = 0.15,
+        min_atoms_per_train: int = 200,
+        max_force_error: float = inf,
+        max_atoms_from_frame: int = inf,
+        max_trains: int = inf,
+        max_model_size: int = inf,
+        max_elts_per_frame: Dict[str, int] = None,
+        max_model_elts: Dict[str, int] = None,
+        predict_atoms_per_elt: Dict[str, int] = None,
+        write_model_train_interval: int = 1,
+        write_model_atom_interval: int = 100,
+        validate_ratio: float = 0,
+        post_write: bool = True,
+    ):
+
+        # Perform pre-run, in which seed trames are used.
+        logger = logging.getLogger(self.logger_name)
+
+        if len(self.gp) == 0:
+            logger.warning(
+                "You are attempting active learning with an empty model. "
+                "One atom of each element will be added from the first frame, "
+                "but be warned: Hyperparameter optimzation on a very small "
+                "subset of data can lead to suboptimal training set "
+                "choices, as the hyperparameters will take time to become "
+                "representative of their converged state relative to your data of "
+                "interest."
+            )
+            self.run_passive_learning(
+                frames[0:1], max_model_elts={elt: 1 for elt in frames[0].species_labels}
+            )
+
+        if isinstance(frames, list):
+            frames = Trajectory(deepcopy(frames))
+
+        train_frame = int(len(frames) * (1 - validate_ratio))
+
+        # Loop through trajectory.
+        train_model_atom_counter = 0  # Track atoms added for training
+        write_model_atom_counter = 0  # Track atoms added for writing
+        train_counter = 0  # Track # of times training done
+
+        # Keep track of which atoms trigger force / uncertainty condition
+        training_plan = {}
+
+        # MAIN LOOP - Frames
+        for i, cur_frame in enumerate(frames):
+            frame_start_time = time.time()
+            logger.info(f"=====NOW ON FRAME {i}=====")
+
+            # If no predict_atoms_per_element was specified, predict_atoms
+            # will be equal to every atom in the frame.
+            predict_atoms = subset_of_frame_by_element(cur_frame, predict_atoms_per_elt)
+
+            # Atoms which are skipped will have NaN as their force / std values
+            local_energies = None
+
+            # Three different predictions: Either MGP, GP with energy,
+            # or GP without
+            if self.mgp:
+                pred_forces, pred_stds, local_energies = self.pred_func(
+                    structure=cur_frame,
+                    mgp=self.gp,
```

Comment:
should this be `self.gp` or `self.mgp`?

Reply:
Ah- `self.mgp` is a flag that indicates if the GP is an `mgp` or not. Should this be renamed to something like `is_mgp`?

Reply:
Ah I see. I thought they represented different modules. Yeah I agree it would be better to rename it like `self.gp_is_mapped` or something similar